pkg/extendapi/devicetemplate.go

```diff
@@ -0,0 +1,196 @@
+		return
+	}
+
+	if err := h.validTemplate(req.Context(), deviceType, templateBytes); err != nil {
+		logrus.Error("CreateDeviceTemplateHealthHandler validTemplate error:", err.Error())
+		w.Write([]byte(fmt.Sprintf(`{"error":%s}`, err.Error())))
+		return
+	}
+
+	if err := h.createSecret(req.Context(), deviceType, user, templateBytes); err != nil {
+		logrus.Error("CreateDeviceTemplateHealthHandler createSecret error:", err.Error())
+		w.Write([]byte(fmt.Sprintf(`{"error":%s}`, err.Error())))
+		return
+	}
+
+	w.Write([]byte(fmt.Sprintf(`{"error":success}`)))
+}
+
+func (h *CreateDeviceTemplateHealthHandler) createSecret(ctx context.Context, deviceType, user string, templateBytes []byte) error {
+
+	r := rand.New(rand.NewSource(time.Now().Unix()))
+	len := 5
+	temStr := make([]byte, len)
+	for i := 0; i < len; i++ {
+		b := r.Intn(26) + 65
+		temStr[i] = byte(b)
+	}
+
+	temp := strings.ToLower(string(temStr))
+	deviceType = strings.ToLower(deviceType)
+
+	name := fmt.Sprintf("%s-template-%s", deviceType, temp)
+	stringData := make(map[string]string)
+	stringData["template"] = string(templateBytes)
+	stringData["user"] = user
+
+	secret := &apiv1.Secret{
+		TypeMeta:   metav1.TypeMeta{Kind: "Secret", APIVersion: "v1"},
+		ObjectMeta: metav1.ObjectMeta{Name: name, Namespace: "octopus-system", Annotations: map[string]string{"user": user}},
+		StringData: stringData,
+	}
+
+	tctx, cancel := context.WithTimeout(ctx, 10*time.Second)
+	defer cancel()
+
+	if _, err := h.clientset.CoreV1().Secrets("octopus-system").Create(tctx, secret, metav1.CreateOptions{}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (h *CreateDeviceTemplateHealthHandler) validTemplate(ctx context.Context, deviceType string, templateBytes []byte) error {
+
+	r := rand.New(rand.NewSource(time.Now().Unix()))
+	len := 5
+	temStr := make([]byte, len)
+	for i := 0; i < len; i++ {
+		b := r.Intn(26) + 65
+		temStr[i] = byte(b)
+	}
+
+	templateMap := make(map[string]interface{})
+
+	js, err := yaml2.YAMLToJSON(templateBytes)
+	if err != nil {
+		return err
+	}
+
+	if err := json.Unmarshal(js, &templateMap); err != nil {
+		return err
+	}
+
+	tmp := strings.ToLower(string(temStr))
+
+	name := fmt.Sprintf("testdevicelink-%s", tmp)
+
+	obj := unstructured.Unstructured{
+		Object: map[string]interface{}{
+			"apiVersion": "edge.cattle.io/v1alpha1",
+			"kind":       "DeviceLink",
+			"metadata": map[string]interface{}{
+				"name":      name,
+				"namespace": "octopus-system",
+			},
+			"spec": map[string]interface{}{
+				"adaptor":  map[string]interface{}{"name": "adaptors.edge.cattle.io/ble", "node": "vm-0-7-ubuntu"},
+				"model":    map[string]interface{}{"apiVersion": "devices.edge.cattle.io/v1alpha1", "kind": "BluetoothDevice"},
+				"template": templateMap,
+			},
+		},
+	}
+
+	opt := metav1.CreateOptions{DryRun: []string{metav1.DryRunAll}}
+
+	resource := schema.GroupVersionResource{Group: "edge.cattle.io",
+		Version:  "v1alpha1",
+		Resource: "devicelinks",
+	}
```

Comment:
the validation should be run against each device CRD spec, e.g, bluetoothdevice, modbusdevices and any custom device spec.